File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int cnt = 0;
    static boolean[] visited;  
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        visited = new boolean[dungeons.length];
        answer = dfs(0, k, dungeons);
        return answer;
    }
    
    public int dfs(int depth, int fatigue, int[][] dungeons) {
        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i] || fatigue < dungeons[i][0]) {
                continue;
            }
            visited[i] = true;
            dfs(depth+1, fatigue - dungeons[i][1], dungeons);
            visited[i] = false;
        }
        cnt = Math.max(cnt, depth);
        return cnt;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
# 최소 필요 피로도, 소모 피로도
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    arr = list(permutations(dungeons,len(dungeons)))
    temp = k
    for perms in arr :
        cnt = 0
        k = temp
        for perm in perms :
            if k >= perm[0] :
                k -= perm[1]
                cnt += 1
            else :
                break
        answer = max(answer,cnt)
        
    return answer